### DISCUSSION TITLE
Base strategies (LwF, Native) do not advance to the next experience after evaluating on the first experience

### DISCUSSION THREAD

**Discussion by JothamWong** [ContinualAI/avalanche#439] [Q&A] [answered]
I am using the standard template from Avalanche's tutorial page "Putting it all together" found here https://avalanche.continualai.org/from-zero-to-hero-tutorial/6.-putting-all-together
I am looping through the scenario's train stream which has 8 experiences
The issue arises when I use one of the base strategies included in the Avalanche framework. The strategies that I have tried to use are LwF and Native. 
What happens is that the strategy will train and evaluate fine, it will print the results to the standard output, and then just pause there. I have included a print statement after the strategy.eval line but it never prints.
I have tried to reproduce this issue with a strategy that I have implemented myself and it works fine, after the eval results are printed, it moves onto the next experience.
I'm hoping someone could help me find out the cause of this issue.

**Comment by Mattdl** [comment]
Could you provide the main script (with the main loop over experiences etc.) so we can have a look? Thanks!

**Comment by JothamWong** [comment, marked as answer]
```import argparse

from avalanche.evaluation.metrics import ExperienceForgetting, accuracy_metrics, loss_metrics, timing_metrics, \
    cpu_usage_metrics, StreamConfusionMatrix, disk_usage_metrics
from avalanche.logging import InteractiveLogger, TextLogger, TensorboardLogger
from avalanche.training.plugins import EvaluationPlugin

import utils


def parse_arguments():
    argparser = argparse.ArgumentParser("Run continual learning experiments")
    argparser.add_argument("--model", type=str)
    argparser.add_argument("--scenario", choices=["NI", "NIC", "NC"])
    argparser.add_argument("--dataset", choices=["MNIST", "CIFAR10", "Core50"])
    argparser.add_argument("--task_list", nargs="+")
    argparser.add_argument("--nb_tasks", type=int)
    argparser.add_argument("--optimizer", type=str, default="SGD")
    argparser.add_argument("--lr", type=float, default=0.001)
    argparser.add_argument("--momentum", type=float, default=0.9)
    argparser.add_argument("--criterion", type=str, default="CrossEntropyLoss")
    argparser.add_argument("--strategy", type=str, choices=["Naive", "DAT", "LwF"])
    argparser.add_argument("--epochs", type=int, default=10)
    argparser.add_argument("--batch_size", type=int, default=16)
    argparser.add_argument("--file_name", type=str, default="log.txt")
    argparser.add_argument("--n_classes", type=int)
    return argparser.parse_args()


print("Parsing arguments...")
args = parse_arguments()
print("Arguments parsed.")
scenario = utils.get_scenario(args)
model = utils.get_model(scenario, args)

tb_logger = TensorboardLogger()
filename = args.file_name
text_logger = TextLogger(open(filename, "a"))
interactive_logger = InteractiveLogger()

eval_plugin = EvaluationPlugin(
    accuracy_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
    loss_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
    timing_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
    cpu_usage_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
    ExperienceForgetting(),
    StreamConfusionMatrix(num_classes=args.n_classes, save_image=True),
    disk_usage_metrics(minibatch=True, epoch=True, experience=True, stream=True),
    loggers=[interactive_logger, text_logger, tb_logger]
)

cl_strategy = utils.get_strategy(model, eval_plugin, args)

print("Starting experiment...")
print(len(scenario.train_stream))

for experience in scenario.train_stream:
    print("Start of experience ", experience.current_experience)
    print("Current classes: ", experience.classes_in_this_experience)

    cl_strategy.train(experience)
    print("Training complete.")

    print("Computing accuracy on the test set")
    cl_strategy.eval(scenario.test_stream)
    print("End of experience ", experience.current_experience)
```

Hi here is the main script. When I use my own custom strategy, it works fine, and the next experience is reached.
When I use any of the base strategies, the results of the test stream is printed to the Interactive Logger and the next print line statement is never executed.

**Reply by JothamWong** [reply]
And this is the script for getting the strategies

```
def get_strategy(model, eval_plugin, args):
    optimizer = args.optimizer
    lr = args.lr
    if optimizer == "SGD":
        momentum = args.momentum
        optimizer = SGD(model.parameters(), lr=lr, momentum=momentum)
    elif optimizer == "Adam":
        optimizer = Adam(model.parameters(), lr=lr)
    else:
        raise ValueError(f"Optimizer {optimizer} currently not supported")

    criterion = args.criterion
    if criterion == "CrossEntropyLoss":
        criterion = CrossEntropyLoss()
    else:
        raise ValueError(f"Criterion {criterion} currently not supported")
    device = torch.device("cuda")
    strategy = args.strategy
    epochs = args.epochs
    batch_size = args.batch_size
    if strategy == "Naive":
        return Naive(
            model=model,
            optimizer=optimizer,
            criterion=criterion,
            train_mb_size=batch_size,
            train_epochs=epochs,
            eval_mb_size=batch_size,
            evaluator=eval_plugin,
            device=device
        )
    elif strategy == "DAT":
        return DomainAdversarialTraining(
            model=model,
            optimizer=optimizer,
            criterion=criterion,
            nb_classes=args.n_classes,
            evaluator=eval_plugin,
            train_mb_size=batch_size,
            eval_mb_size=batch_size,
            train_epochs=epochs,
            device=device
        )
    elif strategy == "LwF":
        return LwF(
            model=model,
            optimizer=optimizer,
            criterion=criterion,
            alpha=1.0,
            temperature=1.0,
            evaluator=eval_plugin,
            train_mb_size=batch_size,
            eval_mb_size=batch_size,
            train_epochs=epochs,
            device=device
        )
    else:
        raise ValueError(f"Strategy {strategy} currently not supported")```

**Reply by Mattdl** [reply]
This looks fine to me, is it possible you have altereded or overwritten some things in the Base Strategy to make your DAT strategy work? In that case, this can give unexpected behaviour to the other methods. Make sure you inherit from the base strategy instead to overwrite. 

This 'pausing' seems weird, have you tried waiting long enough? Could be if the interactive mode is off it takes some time to complete the experience.

**Reply by JothamWong** [reply]
No, I have not made any changes to the code in any Avalanche module. And yes, my DAT strategy does inherit from BaseStrategy class.

**Reply by AndreaCossu** [reply]
Hi @psychicmario ! Can you provide also the `utils.get_scenario` code?

**Reply by JothamWong** [reply]
hey @AndreaCossu, here it is

```
def get_scenario(args):
    train_transform = transforms.Compose([
        transforms.ToTensor()
    ])
    eval_transform = transforms.Compose([
        transforms.ToTensor()
    ])
    if args.dataset == "MNIST" and args.scenario == "NC":
        return SplitMNIST(n_experiences=args.nb_tasks)
    elif args.dataset == "CIFAR10" and args.scenario == "NC":
        return SplitCIFAR10(n_experiences=args.nb_tasks)
    elif args.dataset == "Core50" and args.scenario == "NI":
        return CORe50(root="Datasets/", scenario="ni", train_transform=train_transform, eval_transform=eval_transform)
    elif args.dataset == "Core50" and args.scenario == "nicv2_79":
        return CORe50(root="Datasets/", scenario="nicv2_79", train_transform=train_transform, eval_transform=eval_transform)
    else:
        raise ValueError("Scenario/Dataset currently not supported.")
```

Also @Mattdl you were right. It does eventually go onto the next experience but it takes sometime. I had to wait an hour for it to go to the second experience and then for the next few experiences, it seems to take 30 mins - 1 hour for the next experience to start. I have enabled the Interactive Logger as well, which is how I know if the next experience has started or not. 

Do you have any idea why it takes such a long time to go over to the next experience?

**Reply by Mattdl** [reply]
Good to hear! Depending on using CPU/GPU, the size of your model, and most importantly, which dataset, it can take some time per experience (I also see you defined 10 epochs, so takes 10x longer per experience).

Also LwF might be implemented to first take the network outputs for the new data when transferring to the new experience (not sure it's implemented this way though), so that could take some time in preprocessing of a new experience.

**Reply by JothamWong** [reply]
Hey just an update, when I use the AR1 strategy, there is no delay from experience to experience.
The delay in experience seems to be happening when I use either the Naive or the LwF strategy.

**Reply by AntonioCarta** [reply]
1 hour is not a long time  but it seems longer than I would expect from your settings. Can you provide a self-contained script to reproduce the issue? If you provide a single, minimal script with a single strategy and scenario we may be able to investigate the issue.
It would also be useful to know if you are using GPUs or CPUs and your operating system.

**Reply by AndreaCossu** [reply]
@psychicmario just to make sure we intend the same thing. When you say it took 30min - 1hr to go to the second experience do you mean that training on the first experience (+ eval) took 1hr **or** that after training on the first experience (+eval) you had to wait 1hr before the training on second experience begins?
The former could be normal, as @Mattdl and @AntonioCarta said. The latter however would be very strange.

**Reply by JothamWong** [reply]
@AndreaCossu 
I am using my company's server to do the training which is quite fast. Looking at the most recent standard output, training and evaluating on the Core50 NI scenario took 2.5 mins. It then took an hour to go to the second experience. So to answer your question directly, after training on the first experience, I had to wait between 30 mins and 1 hour for training on the second experience to begin.

**Reply by JothamWong** [reply]
@AntonioCarta 
Here is a sample script which I can replicate the issue on 100% of the time
```import torch
import torch.optim as optim
import torchvision.transforms as transforms
from avalanche.benchmarks.classic import CORe50
from avalanche.evaluation.metrics import ExperienceForgetting, accuracy_metrics, loss_metrics, timing_metrics, \
    cpu_usage_metrics, StreamConfusionMatrix, disk_usage_metrics
from avalanche.logging import InteractiveLogger, TextLogger, TensorboardLogger
from avalanche.training.plugins import EvaluationPlugin
from avalanche.training.strategies import Naive
from torchvision.models import resnet34

# Average results over 5 runs
nb_experiments = 5
for experiment in range(nb_experiments):
    train_transform = transforms.Compose([
        transforms.ToTensor()
    ])
    eval_transform = transforms.Compose([
        transforms.ToTensor()
    ])
    scenario = CORe50(root="Datasets/", scenario="ni", train_transform=train_transform, eval_transform=eval_transform)

    model = resnet34(pretrained=True)

    name = f"experiment {experiment+1}"
    tb_logger = TensorboardLogger(tb_log_exp_name=name)
    filename = name + ".txt"
    text_logger = TextLogger(open(filename, "a"))
    interactive_logger = InteractiveLogger()

    eval_plugin = EvaluationPlugin(
        accuracy_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
        loss_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
        timing_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
        cpu_usage_metrics(minibatch=True, epoch=True, epoch_running=True, experience=True, stream=True),
        ExperienceForgetting(),
        StreamConfusionMatrix(num_classes=50, save_image=True),
        disk_usage_metrics(minibatch=True, epoch=True, experience=True, stream=True),
        loggers=[interactive_logger, text_logger, tb_logger]
    )
    optimizer = optim.SGD(model.parameters(), lr=1e-3, momentum=0.9)
    criterion = torch.nn.CrossEntropyLoss()
    batch_size = 16
    epochs = 5
    device = torch.device("cuda")

    cl_strategy = Naive(
            model=model,
            optimizer=optimizer,
            criterion=criterion,
            train_mb_size=batch_size,
            train_epochs=epochs,
            eval_mb_size=batch_size,
            evaluator=eval_plugin,
            device=device
        )

    print(f"Starting experiment {experiment + 1}...")
    print(len(scenario.train_stream))

    for experience in scenario.train_stream:
        print("Start of experience ", experience.current_experience)
        print("Current classes: ", experience.classes_in_this_experience)

        cl_strategy.train(experience)
        print("Training complete.")

        print("Computing accuracy on the test set")
        cl_strategy.eval(scenario.test_stream)
        print("End of experience ", experience.current_experience)```
I am using a GPU, although I am unsure what model it is exactly, and my operating system is Linux
For the script above, it seems to "pause" at the eval line for awhile before the next print statement is executed and it moves onto the next experience.

**Reply by JothamWong** [reply]
Just to add, I had also forced a reinstall of the library just in case I had accidentally modified the source code or something and tried again on the script above but to no luck. My custom strategy along with the AR1 strategy works fine however.

**Reply by vlomonaco** [reply]
@psychicmario sorry for the late answer! Have you been able to fix the issue? I will try to debug it asap.

**Reply by JothamWong** [reply]
@vlomonaco No I havent been able to sadly. This issue still arises for Naive/LwF and sometimes happens for my custom strategy.

**Reply by vlomonaco** [reply]
Hi @psychicmario, it seems the problem is related to the `StreamConfusionMatrix` metric, can you try re-run your script commenting out that metric? On my Linux server with a fresh env it doesn't stall anymore.

**Reply by AndreaCossu** [reply]
Hi @psychicmario ! We are working on solving the issue, it should not take long. You can track the advancements here #513

**Reply by AndreaCossu** [reply]
#518 should have solved this. Please, @psychicmario try again and see if it works now. Check out the class documentation of `ConfusionMatrix` and `StreamConfusionMatrix` for more details.

**Reply by vlomonaco** [reply]
Marking this as solved due to inactivity, @psychicmario let us know in case you want to re-open it.

**Reply by JothamWong** [reply]
Hey, sorry for the late reply. I just tested my script with the ConfusionMatrix metric and I was able to get it training without a delay. Thanks @AndreaCossu, @vlomonaco for solving the problem :)
